### ISSUE TITLE
USART serial availableForWrite() not working

### ISSUE THREAD

**Issue by rickkas7** [spark/firmware#1017]
There’s a simple typo in the implementation of `USARTSerial::availableForWrite()`:

```cpp
int USARTSerial::availableForWrite(void)
{
return HAL_USART_Available_Data(_serial);
}
int USARTSerial::available(void)
{
return HAL_USART_Available_Data(_serial);
}
```

https://github.com/spark/firmware/blob/develop/wiring/src/spark_wiring_usartserial.cpp#L69

It should probably be `HAL_USART_Available_Data_For_Write`. However, if I fix that and use it with code like:

```cpp
        Serial.printlnf("%d", Serial1.availableForWrite());
        Serial1.println("01234567890123456789012345678901234567890123456789");
        Serial.printlnf("%d", Serial1.availableForWrite());
```

The output is:

```
0
50
```

I’m unclear of how the function is supposed to work. Should it:

1. Return the number of bytes waiting to be written out from the buffer (0=empty send buffer), which is what it does.
2. Return the number of bytes you can write to the buffer without blocking, which would be SERIAL_BUFFER_SIZE - the value currently returned.

Thanks,
Rick

---

Completeness:
- [ ] Minimum test case added
- [ ] Device, system and user firmware versions stated
- [x] Particle confirmed

**Comment by m-mcgowan** [comment]
@avtolstoy Any thoughts on this?

**Comment by avtolstoy** [comment]
Confirmed and fixed in #1020 :
1. A typo in `USARTSerial::availableForWrite()`
2. `availableForWrite()` behavior, which should behave as described in docs:
> Retrieves the number of bytes (characters) that can be written to this serial port without blocking.

**Comment by rickkas7** [comment]
Thanks! Sorry, I swear I looked it up in the docs and didn’t see it, but yes, the correct behavior is documented.
